program/rust/src/time_machine_types.rs

```diff
@@ -6,20 +6,353 @@ use crate::c_oracle_header::{
+            / try_convert::<u64, SignedTrackerRunningSum>(self.current_entry_slot_accumulator)?;
+        //compute the current entry's confidence
+        self.running_entry_confidences[entry] = self.running_entry_confidences[prev_entry];
+        self.running_entry_confidences[entry] += self.current_entry_weighted_confidence_accumulator
+            / try_convert::<u64, UnsignedTrackerRunningSum>(self.current_entry_slot_accumulator)?;
+        //check current entry validity
+        self.running_valid_entry_counter[entry] =
+            if self.current_entry_status == Status::Pending && slot_gap < self.threshold {
+                self.running_valid_entry_counter[prev_entry] + 1
+            } else {
+                self.running_valid_entry_counter[prev_entry]
+            };
+
+        //setup working variables for the next entry
+        self.current_entry_slot_accumulator = 0;
+        self.current_entry_weighted_confidence_accumulator = 0;
+        self.current_entry_weighted_price_accumulator = 0;
+        self.current_entry_status = if slot_gap < self.threshold {
+            Status::Pending
+        } else {
+            Status::Invalid
+        };
+        Ok(())
+    }
+}
+
+
+#[derive(Debug, Copy, Clone)]
+#[repr(C)]
+/// A Tracker that has NUM_ENTRIES entries
+/// each tracking the last tick (price and confidence) before every GRANULARITY seconds.
+pub struct TickTracker<const NUM_ENTRIES: usize> {
+    threshold:      i64,
+    granularity:    i64,
+    prices:         [SignedTrackerRunningSum; NUM_ENTRIES], /* price running time
+                                                             * weighted sums */
+    confidences:    [UnsignedTrackerRunningSum; NUM_ENTRIES], /* confidence running
+                                                               * time
+                                                               * weighted sums */
+    entry_validity: [Status; NUM_ENTRIES], /* entry validity, INVALID if the last tick in an
+                                            * entry is too far from the end, PENDING if it is
+                                            * the entry is still being updated but the current
+                                            * tick is close enough, VALID if the entry is no
+                                            * longer being updated and the last tick is
+                                            * sufficiently close */
+}
+
+
+impl<const NUM_ENTRIES: usize> TickTracker<NUM_ENTRIES> {
+    fn initialize(&mut self, threshold: i64, granularity: i64) -> Result<(), OracleError> {
+        self.threshold = threshold;
+        self.granularity = granularity;
+        Ok(())
+    }
+
+    ///add a new price to the tracker
+    fn add_price(
+        &mut self,
+        prev_time: i64,
+        current_time: i64,
+        current_price: i64,
+        current_conf: u64,
+    ) -> Result<(), OracleError> {
+        self.close_previous_tick(prev_time, current_time)?;
```

Comment:
fwiw i think this code would be clearer if you got rid of these inner functions, as it would be like 50% of the length. (of course, there will be other times where that will be underabstracting, but this feels a little overabstracted to me)